tiledb/sm/array/consistency.h

```diff
@@ -0,0 +1,195 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines classes ConsistencyController and ConsistencySentry.
+ */
+
+#ifndef TILEDB_CONSISTENCY_H
+#define TILEDB_CONSISTENCY_H
+
+#include <iterator>
+#include <map>
+
+#include "tiledb/common/common.h"
+#include "tiledb/common/logger.h"
+#include "tiledb/sm/filesystem/uri.h"
+
+using namespace tiledb::common;
+
+namespace tiledb::sm {
+
+class Array;
+class ConsistencySentry;
+
+/**
+ * Tracks the open arrays on disk, considering that a given
+ * URI can have multiple open arrays.
+ *
+ * Intended to act as a singleton global.
+ * There is only one (global) copy of this class used in practice.
+ */
+class ConsistencyController {
+  /**
+   * Makes available the private APIs of ConsistencyController for array
+   * registration and deregistration.
+   */
+  friend class ConsistencySentry;
+  /**
+   * Makes available internals of ConsistencyController for testing.
+   */
+  friend class WhiteboxConsistencyController;
+
+ public:
+  using entry_type = std::multimap<const URI, Array&>::const_iterator;
+
+  /**
+   * Constructor.
+   *
+   * Default constructed as the initial state of the singleton Controller with
+   * no registered arrays.
+   * Note: nothing can be registered in the singleton until an array is opened.
+   */
+  ConsistencyController() = default;
+
+  /** Copy Constructor is deleted. */
+  ConsistencyController(const ConsistencyController&) = delete;
+
+  /** Move Constructor is deleted. */
+  ConsistencyController(ConsistencyController&&) = delete;
+
+  /** Copy assignment is deleted. */
+  ConsistencyController& operator=(const ConsistencyController&) = delete;
+
+  /** Move assignment is deleted. */
+  ConsistencyController& operator=(ConsistencyController&&) = delete;
+
+  /** Destructor. */
+  ~ConsistencyController() = default;
+
+  /**
+   * Register the given Array as open.
```

Comment:
We (Eric, Isaiah, and I) have all discussed this at one point. I'm still open to changing the nomenclature. However, a user should not be calling `make_sentry`. The only place this is ever called (outside of the Whitebox) is in `Array::set_array_open()`, which is a private method. As far as a user can tell, the Sentry represents "Lifespan maintenance of an open array in the ConsistencyController." So calling `make_sentry` inside of `set_array_open` makes more sense from the perspective of the Array class. 
But I also understand where you're coming from, we don't want it to be confusing for users. If @ihnorton or @eric-hughes-tiledb has a strong opinion one way or another, please let me know. 
(Note the internal `ConsistencyController` function is called `register_array` - this would also need to be renamed.)

Reply:
> my intention is to register an array

You accomplish that task by creating an object whose lifespan matches the duration of array being open, not by simply calling a function on the controller itself. The sentry is designed to be a private member variable of a resource class. It's not quite that today. The member variable is declared `optional<ConsistencySentry>` because `class Array` is not a resource class.

Using a sentry class eliminates any special-case code about deregistering, regardless of circumstance.